File: reb_3_4/reb_3_4_calculations.py
import pandas as pd
from scipy import optimize

# Given or known
n_CO_0 = 34.0 # mol
n_H2_0 = 66.0 # mol
n_CH3OH_0 = 0.0
n_CH4_0 = 0.0
n_H2O_0 = 0.0
n_CO2_0 = 0.0
f_CO = 0.4
Y_CH3OH = 0.107
S_CH3OH_CH4 = 0.38

# Solve eqns 7 through 9 for the apparent extents of reaction
# equations to solve, as residuals
def residuals(x):
    return [x[0] + x[1] + x[2] - n_CO_0*f_CO,
        n_CH3OH_0 + x[0] - Y_CH3OH*n_CO_0,
        n_CH3OH_0 + x[0] - S_CH3OH_CH4*(n_CH4_0 + x[1])]
# guess for solution
guess = [10, 10, 10]
# try to solve
soln = optimize.root(residuals, guess)
# check for success
if not(soln.success):
    print("A solution was NOT obtained:")
    print(soln.message)
# extract the solution
extent = soln.x

# Calculate the final moles using eqns 10 through 15
n_CO = n_CO_0 - extent[0] - extent[1] - extent[2]
n_H2 = n_H2_0 - 2*extent[0] - 3*extent[1] + extent[2]
n_CH3OH = n_CH3OH_0 + extent[0]
n_CH4 = n_CH4_0 + extent[1]
n_H2O = n_H2O_0 + extent[1] - extent[2]
n_CO2 = n_CO2_0 + extent[2]

# Display the result
print(' ')
print('Final Moles')
print(' CO:','%.3g' % n_CO)
print(' H2:','%.3g' % n_H2)
print(' CH3OH:','%.3g' % n_CH3OH)
print(' CH4:','%.3g' % n_CH4)
print(' H2O:','%.3g' % n_H2O)
print(' CO2:','%.3g' % n_CO2)
print(' ')

# Save the result to a .csv file
data = [['n_CO', '%.3g' % n_CO, 'mol'], \
    ['n_H2', '%.3g' % n_H2, 'mol'], \
    ['n_CH3OH', '%.3g' % n_CH3OH, 'mol'], \
    ['n_CH4', '%.3g' % n_CH4, 'mol'], \
    ['n_H2O', '%.3g' % n_H2O, 'mol'], \
    ['n_CO2', '%.3g' % n_CO2, 'mol'], \
]
result = pd.DataFrame(data, columns=['item','value','units'])
result.to_csv("./reb_3_4/reb_3_4_Python_results.csv", index=False)

File: reb_3_4/Code/reb_3_4_calculations.m
function reb_3_4_calculations
    % Given or known
    n_CO_0 = 34.0; % mol
    n_H2_0 = 66.0; % mol
    n_CH3OH_0 = 0;
    n_CH4_0 = 0;
    n_H2O_0 = 0;
    n_CO2_0 = 0;
    f_CO = 0.4;
    Y_CH3OH = 0.107;
    S_CH3OH_CH4 = 0.38;

    % Solve eqns 7 through 9 for the apparent extents of reaction
    % equations to solve, as residuals
    function residual = calcResid(x_test)
        residual = [
            f_CO*n_CO_0 - x_test(1) - x_test(2) - x_test(3)
            Y_CH3OH*n_CO_0 - n_CH3OH_0 - x_test(1)
            S_CH3OH_CH4 *(n_CH4_0 + x_test(2)) - n_CH3OH_0 - x_test(1)
        ];
    end
    % guess for the solution
    x_guess = [10, 10, 10];
    % try to solve
    [x, flag, message] = solveATEs(@calcResid, x_guess);
    % check for success
    if flag <= 0
        disp('The equations were NOT solved.')
        disp(['    Matlab provided the following message: ',message])
    end

    % Calculate the molar amounts, eqns 10 through 15
    n_CO = n_CO_0 - x(1) - x(2) - x(3);
    n_H2 = n_H2_0 - 2*x(1) - 3*x(2) + x(3);
    n_CH3OH = n_CH3OH_0 + x(1);
    n_CH4 = n_CH4_0 + x(2);
    n_H2O = n_H2O_0 + x(2) - x(3);
    n_CO2 = n_CO2_0 + x(3);

    % Display the result
    disp(' ')
    disp('Final Moles')
    disp([' CO: ', num2str(n_CO,3)])
    disp([' H2: ', num2str(n_H2,3)])
    disp([' CH3OH: ', num2str(n_CH3OH,3)])
    disp([' CH4: ', num2str(n_CH4,3)])
    disp([' H2O: ', num2str(n_H2O,3)])
    disp([' CO2: ', num2str(n_CO2,3)])

    % Save the result to a .csv file
    results_file ="reb_3_4_Matlab_results.csv";
    item = ["n_CO";"n_H2";"n_CH3OH";"n_CH4";"n_H2O";"n_CO2"];
    value = [round(n_CO,3,'significant'); round(n_H2,3,'significant'); ...
        round(n_CH3OH,3,'significant'); round(n_CH4,3,'significant'); ...
        round(n_H2O,3,'significant'); round(n_CO2,3,'significant')];
    units = ["mol";"mol";"mol";"mol";"mol";"mol"];
    results_table = table(item,value,units);
    writetable(results_table,results_file);
end